python/tests/test_topology.py

```diff
@@ -904,6 +903,400 @@ def test_internal_samples(self):
             t1.kc_distance(t2)
 
 
+def ts_kc_distance(ts1, ts2, lambda_=0):
+    check_kc_tree_sequence_inputs(ts1, ts2)
+
+    total = 0
+    left = 0
+    tree1_iter = ts1.trees(sample_lists=True)
+    tree1 = next(tree1_iter)
+    for tree2 in ts2.trees(sample_lists=True):
+        while tree1.interval[1] < tree2.interval[1]:
+            span = tree1.interval[1] - left
+            total += c_kc_distance(tree1, tree2, lambda_) * span
+
+            left = tree1.interval[1]
+            tree1 = next(tree1_iter)
+        span = tree2.interval[1] - left
+        left = tree2.interval[1]
+        total += c_kc_distance(tree1, tree2, lambda_) * span
+
+    return total / ts1.sequence_length
+
+
+def ts_kc_distance_incremental(ts1, ts2, lambda_=0):
+    check_kc_tree_sequence_inputs(ts1, ts2)
+
+    sample_maps = [dict(), dict()]
+    for i, ts in enumerate([ts1, ts2]):
+        for j, u in enumerate(ts.samples()):
+            sample_maps[i][u] = j
+
+    total = 0
+    left = 0
+
+    t1_vecs = KCVectors(ts1.num_samples)
+    t2_vecs = KCVectors(ts2.num_samples)
+
+    edge_diffs_iter_1 = ts1.edge_diffs()
+    tree_iter_1 = ts1.trees(sample_lists=True)
+    t1, t1_diffs = next(tree_iter_1), next(edge_diffs_iter_1)
+    update_kc_incremental(t1, t1_vecs, t1_diffs, sample_maps[0])
+    for t2, t2_diffs in zip(ts2.trees(sample_lists=True), ts2.edge_diffs()):
+        update_kc_incremental(t2, t2_vecs, t2_diffs, sample_maps[1])
+        while t1_diffs[0][1] < t2_diffs[0][1]:
+            span = t1_diffs[0][1] - left
+            total += norm_kc_vectors(t1_vecs, t2_vecs, lambda_) * span
+
+            left = t1_diffs[0][1]
+            t1, t1_diffs = next(tree_iter_1), next(edge_diffs_iter_1)
+            update_kc_incremental(t1, t1_vecs, t1_diffs, sample_maps[0])
+        span = t2_diffs[0][1] - left
+        left = t2_diffs[0][1]
+        total += norm_kc_vectors(t1_vecs, t2_vecs, lambda_) * span
+
+    return total / ts1.sequence_length
+
+
+def update_kc_incremental(tree, kc, edge_diffs, sample_index_map):
+    _, edges_out, edges_in = edge_diffs
+
+    # Update any depths/times changed by removed edges that might not be updated
+    # by inserted edges.
+    for e in edges_out:
+        update_in_subtree_pairs(tree, kc, e.child, sample_index_map)
+
+    for e in edges_in:
```

Comment:
I think this is the source of the initial slow down. For edges `e1` and `e2` in `edges_in`, if `e1` is a descendant of `e2`, then the set of pairs associated with `e1` is a subset of those associated with `e2`. The updates completed when processing `e1` would then get completely overwritten by those of `e2`. This is especially egregious in the case of the first tree, where this is doing `O(E * n^2)` updates. It should be enough (in the first tree's case) to only process those edges directly under the root, and doing so should run essentially the same algorithm as the single tree kc. This would then generalize to processing only those edges without another incoming edge on the path from that edge to the root.

Reply:
Decided that this should be a performance optimization in a second pass, not in this PR.